Samples/WinMLSamplesGallery/TransformAsync/TransformAsync.cpp

```diff
@@ -0,0 +1,934 @@
+    {
+        delete this;
+    }
+    // For thread safety, return a temporary variable.
+    return count;
+}
+#pragma endregion IUnknown
+
+
+HRESULT TransformAsync::ScheduleFrameInference(void)
+{
+    com_ptr<IMFSample> inputSample; 
+    com_ptr<IMFAsyncResult> result;
+
+    // No need to lock, sample queues are thread safe
+     /***************************************
+        ** Since this in an internal function
+        ** we know m_inputSampleQueue can never be
+        ** NULL due to InitializeTransform()
+        ***************************************/
+
+    RETURN_IF_FAILED(m_inputSampleQueue->GetNextSample(inputSample.put()));
+    
+    RETURN_IF_FAILED(MFCreateAsyncResult(inputSample.get(),   // Pointer to object stored in async result
+        this,                               // Pointer to IMFAsyncCallback interface
+        this,                               // Pointer to IUnkonwn of state object
+        result.put()));
+
+
+    // Schedule the inference task with the MF async work queue. Will call TransformAsync::Invoke when run.
+    RETURN_IF_FAILED(MFPutWorkItemEx(MFASYNC_CALLBACK_QUEUE_MULTITHREADED, result.get()));
+    return S_OK;
+}
+
+// SubmitEval returns an HRESULT to use to signal the async result's status
+HRESULT TransformAsync::SubmitEval(IMFSample* pInput)
+{
+    com_ptr<IMFSample> outputSample;
+    com_ptr<IMFSample> inputSample;
+    IDirect3DSurface src, dest;
+    inputSample.copy_from(pInput); 
+
+    // Select the next available model
+    DWORD dwCurrentSample = InterlockedIncrement(&m_sampleCounter);
+    m_modelIndex = (++m_modelIndex) % m_numThreads;
+    auto model = m_models[m_modelIndex].get();
+
+    //inputSample attributes to copy over to outputSample
+    LONGLONG duration = 0;
+    LONGLONG time = 0;
+    UINT64 markerID = 0;
+
+    TRACE((L"\n[Sample: %d | model: %d | ", currentSample, m_modelIndex));
+    TRACE((L" | SE Thread %d | ", std::hash<std::thread::id>()(std::this_thread::get_id())));
+
+     // Ensure we still have a valid d3d device
+    RETURN_IF_FAILED(CheckDX11Device());
+    // Ensure the allocator is set up 
+    RETURN_IF_FAILED(SetupAlloc());
+
+    // CheckDX11Device & SetupAlloc ensure we can now allocate a D3D-backed output sample. 
+    HRESULT hr = m_outputSampleAllocator->AllocateSample(outputSample.put());
+    if (FAILED(hr))
+    {
+        TRACE((L"Can't allocate any more samples, DROP"));
+    }
+
+    // We should have a sample now, whether or not we have a dx device
+    RETURN_IF_NULL_ALLOC(outputSample);
+
+    // Explicitly copy out pInput attributes, since copying IMFSamples is shallow. 
+    RETURN_IF_FAILED(inputSample->GetSampleDuration(&duration));
+    RETURN_IF_FAILED(inputSample->GetSampleTime(&time));
+    inputSample->GetUINT64(TransformAsync_MFSampleExtension_Marker, &markerID);
+
+    // Extract an IDirect3DSurface from the input and output samples to use for inference, 
+    src = SampleToD3Dsurface(inputSample.get());
```

Comment:
>src [](http://example.com/codeflow?start=4&length=3)

[minor] Propose moving `IDirect3DSurface` down closer to first use (that way one doesn't have to scroll back up so far to see the type, and it keeps relevant scope narrower).